src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Structs.cs

```diff
@@ -0,0 +1,164 @@
+            public TypeWithAnnotations LValueType => VisitResult.LValueType;
+
+            public VisitArgumentResult(VisitResult visitResult, Optional<LocalState> stateForLambda)
+            {
+                VisitResult = visitResult;
+                StateForLambda = stateForLambda;
+            }
+        }
+
+        /// <summary>
+        /// Contains a checkpoint for the NullableWalker at any given point of execution, used for restoring the walker to
+        /// a specific point for speculatively analyzing a piece of code that does not appear in the original tree.
+        /// </summary>
+        internal readonly struct Checkpoint
+        {
+            private readonly VariableState _variableState;
+            private readonly bool _useMethodSignatureParameterTypes;
+            private readonly Symbol _symbol;
+            private readonly ImmutableArray<(BoundReturnStatement, TypeWithAnnotations)> _returnTypesOpt;
+            private readonly ImmutableDictionary<BoundExpression, TypeWithState> _methodGroupReceiverMapOpt;
+            private readonly VisitResult _visitResult;
+            private readonly VisitResult _currentConditionalReceiverVisitResult;
+            private readonly ImmutableDictionary<object, PlaceholderLocal> _placeholderLocalsOpt;
+            private readonly int _lastConditionalAccessSlot;
+
+            private Checkpoint(
+                VariableState variableState,
+                bool useMethodSignatureParameterTypes,
+                Symbol symbol,
+                ImmutableArray<(BoundReturnStatement, TypeWithAnnotations)> returnTypesOpt,
+                ImmutableDictionary<BoundExpression, TypeWithState> methodGroupReceiverMapOpt,
+                VisitResult visitResult,
+                VisitResult currentConditionalReceiverVisitResult,
+                ImmutableDictionary<object, PlaceholderLocal> placeholderLocalsOpt,
+                int lastConditionalAccessSlot)
+            {
+                _variableState = variableState;
+                _useMethodSignatureParameterTypes = useMethodSignatureParameterTypes;
+                _symbol = symbol;
+                _returnTypesOpt = returnTypesOpt;
+                _methodGroupReceiverMapOpt = methodGroupReceiverMapOpt;
+                _visitResult = visitResult;
+                _currentConditionalReceiverVisitResult = currentConditionalReceiverVisitResult;
+                _placeholderLocalsOpt = placeholderLocalsOpt;
+                _lastConditionalAccessSlot = lastConditionalAccessSlot;
+            }
+
+            internal static Checkpoint CheckpointWalker(NullableWalker walker) =>
+                new Checkpoint(
+                    walker.GetVariableState(walker.State),
+                    walker._useMethodSignatureParameterTypes,
+                    walker._symbol,
+                    walker._returnTypesOpt?.ToImmutable() ?? default,
+                    walker._methodGroupReceiverMapOpt?.ToImmutableDictionary(),
```

Comment:
>walker._methodGroupReceiverMapOpt?.ToImmutableDictionary(), [](start = 20, length = 59)

Can we share this `ImmutableDictionary<,>` with all checkpoints, until the original map changes? Same question for other immutable snapshots here.

Reply:
If we're creating checkpoints at each node, then perhaps the actual `NullableWalker` fields should be immutable objects so the checkpoints can simply copy the fields by reference.

---
In reply to: [286220737](https://github.com/dotnet/roslyn/pull/35850#discussion_r286220737) [](ancestors = 286220737)

Reply:
This will force us to create allocations for these new objects in the general straight-line case, which is a tradeoff we can make but I want to be sure we're deliberate about choosing to do so.

---
In reply to: [286227877](https://github.com/dotnet/roslyn/pull/35850#discussion_r286227877) [](ancestors = 286227877,286220737)